### ISSUE TITLE
dd-agent 5.2.2 put critical services down due to having stalled jmx_fetch threads

### ISSUE THREAD

**Issue by ssbarnea** [DataDog/dd-agent#1529]
I reported the some problem months ago and it seems that it was not solved by new versions of dd-agent. Last night datadog put one critical application down due to lack of memory on the server and also failed to send any notifications, even if the service was running.

Before looking at the log I should mention that I restarted the service, but this did nothing because restarting the services did nothing to the jmxfetch running copies.

I will open a support ticket for sending the log and config files.
```
root@atlas:~# ps -Af|grep dd-agent
dd-agent  2103     1  0 Feb09 ?        02:58:35 java -classpath /opt/datadog-agent/agent/checks/libs/jmxfetch-0.4.0-jar-with-dependencies.jar org.datadog.jmxfetch.App --check tomcat.yaml --check_period 15000 --conf_directory /etc/dd-agent/conf.d --log_level WARN --log_location /var/log/datadog/jmxfetch.log --reporter statsd:8125 --status_location /tmp/jmx_status.yaml collect
dd-agent 20961     1  0 Feb26 ?        02:09:25 java -classpath /opt/datadog-agent/agent/checks/libs/jmxfetch-0.4.0-jar-with-dependencies.jar org.datadog.jmxfetch.App --check tomcat.yaml --check_period 15000 --conf_directory /etc/dd-agent/conf.d --log_level WARN --log_location /var/log/datadog/jmxfetch.log --reporter statsd:8125 --status_location /tmp/jmx_status.yaml collect
root     37534 33630  0 07:28 pts/17   00:00:00 grep --color=auto dd-agent
dd-agent 38776     1  0 Mar24 ?        00:56:34 java -classpath /opt/datadog-agent/agent/checks/libs/jmxfetch-0.4.1-jar-with-dependencies.jar org.datadog.jmxfetch.App --check tomcat.yaml --check_period 15000 --conf_directory /etc/dd-agent/conf.d --log_level WARN --log_location /var/log/datadog/jmxfetch.log --reporter statsd:8125 --status_location /tmp/jmx_status.yaml collect
dd-agent 46684     1  0 Mar26 ?        00:36:31 java -classpath /opt/datadog-agent/agent/checks/libs/jmxfetch-0.4.1-jar-with-dependencies.jar org.datadog.jmxfetch.App --check tomcat.yaml --check_period 15000 --conf_directory /etc/dd-agent/conf.d --log_level WARN --log_location /var/log/datadog/jmxfetch.log --reporter statsd:8125 --status_location /tmp/jmx_status.yaml collect
root@atlas:~# free
             total       used       free     shared    buffers     cached
Mem:      49413424   49052904     360520   11777384      18884   19820360
-/+ buffers/cache:   29213660   20199764
Swap:     10485752    1454932    9030820
root@atlas:~# kill 2103 38776 46684 20961
root@atlas:~# free
             total       used       free     shared    buffers     cached
Mem:      49413424   45931988    3481436   11777300      19844   19818620
-/+ buffers/cache:   26093524   23319900
Swap:     10485752    1195908    9289844
root@atlas:~#
```

**Comment by LeoCavaille** [comment]
@ssbarnea thanks for reporting this and sorry for the inconvenience.
In the upcoming 5.3 release, we changed the way the jmxfetch process is supervised, making it more reliable. It looks like that 2 processes were launched at a 1-month interval and with different versions of jmxfetch (0.4.0/0.4.1). So I am betting something went wrong during the upgrade of the package.
I'll have a look to see what we can guess from your logs/configs,  and also we'll run some more tests on the upgrade process.

**Comment by ssbarnea** [comment]
Considering that datadog is running under its own username the service restart should kill! all processes executed by this user, preventing outages like the one presented above.

I see no reason why after a `service datadog-agent stop` I would have to see other processes run under this user.

**Comment by remh** [comment]
@ssbarnea we released agent 5.3.0 yesterday which makes jmxfetch runs on its own in the supervisor.

Could you stop the agent, make sure that no Agent process is running, upgrade and try to reproduce the issue please ?

Thanks for the feedback!

**Comment by ssbarnea** [comment]
I was able to reproduce the problem and I do think that I know what happens. Your deb upgrader does not shutdown dd-agent properly. I performed the upgrade using "apt-get upgrade", as that's the standard way to update any package. 
```
root@atlas:~# ps -Af|grep dd-agent
dd-agent  4782     1  0 Apr13 ?        00:13:27 java -classpath /opt/datadog-agent/agent/checks/libs/jmxfetch-0.4.1-jar-with-dependencies.jar org.datadog.jmxfetch.App --check tomcat.yaml --check_period 15000 --conf_directory /etc/dd-agent/conf.d --log_level WARN --log_location /var/log/datadog/jmxfetch.log --reporter statsd:8125 --status_location /tmp/jmx_status.yaml collect
root     21754 21682  0 17:04 pts/5    00:00:00 grep --color=auto dd-agent
root     35443     1  0 12:36 ?        00:00:03 /opt/datadog-agent/embedded/bin/python /opt/datadog-agent/bin/supervisord -c /etc/dd-agent/supervisor.conf --pidfile /var/run/datadog-supervisord.pid
dd-agent 35447 35443  0 12:36 ?        00:00:36 /opt/datadog-agent/embedded/bin/python /opt/datadog-agent/agent/dogstatsd.py --use-local-forwarder
dd-agent 35448 35443  0 12:36 ?        00:00:39 /opt/datadog-agent/embedded/bin/python /opt/datadog-agent/agent/ddagent.py
dd-agent 35449 35443  2 12:36 ?        00:06:50 /opt/datadog-agent/embedded/bin/python /opt/datadog-agent/agent/agent.py foreground --use-local-forwarder
root     35469 35450  0 12:36 ?        00:00:58 java -Xms50m -Xmx200m -classpath /opt/datadog-agent/agent/checks/libs/jmxfetch-0.5.2-jar-with-dependencies.jar org.datadog.jmxfetch.App --check tomcat.yaml --check_period 15000 --conf_directory /etc/dd-agent/conf.d --log_level WARN --log_location /var/log/datadog/jmxfetch.log --reporter statsd:8125 --status_location /tmp/jmx_status.yaml collect
```
As you can see it already has more than one jmx running.

**Comment by remh** [comment]
That was fixed by agent 5.3.0, closing.
